rfc/rfc-63/rfc-63.md

```diff
@@ -0,0 +1,418 @@
+  However, it is not a pre-requisite for the first cut of functional indexes. Expression engine should be discussed in
+  another RFC.
+
+## Design and Implementation
+
+At a high level, **Functional Index** design principles are as follows:
+
+1. User specifies the Functional Index, including the original data column, the expression applying a function on the
+   column(s), through **SQL** or Hudi's write config. Indexes can be created or dropped for Hudi tables at any time.
+2. While table properties will be the source of truth about what indexes are available, the metadata about each
+   functional index is registered in a separate index metadata file, to keep track of the relationship between the data
+   column and the functional Index. 
+3. Each functional index will be a partition inside the Hudi metadata table (MT).
+4. No data is materialized to the data files for the functional index, i.e., the data files do not contain a new data
+   column corresponding to the Functional Index, to save storage space.
+5. When query engine makes a logical plan for a query, Hudi intercepts the predicate that relates to a functional index,
+   looks for the corresponding index in the MT, and applies data skipping based on one or more indexes.
+6. In order to be engine-agnostic, the design should not make assumptions about any particular engine.
+
+### Components
+
+We discuss the design and implementation details of each component to realize the Functional Index.
+
+#### SQL
+
+A new keyword `FUNCTION` is introduced to indicate that a functional index is intended to be created. This keyword is
+ANSI-compliant and also present in [spark-sql](https://spark.apache.org/docs/latest/sql-ref-ansi-compliance.html).
+
+```sql
+-- PROPOSED SYNTAX WITH FUNCTION KEYWORD --
+CREATE
+[FUNCTION] INDEX index_name ON table_name [ USING index_type ] ( { column_name | expression } );
+-- Examples --
+CREATE FUNCTION INDEX last_name_idx ON employees (UPPER(last_name)); -- functional index using column stats for UPPER(last_name)
+CREATE FUNCTION INDEX datestr ON hudi_table (DATE_FORMAT(timestamp, '%Y-%m-%d')); -- functional index using column stats for DATE_FORMAT(timestamp, '%Y-%m-%d')
+CREATE INDEX city_id_idx ON hudi_table (city_id); -- usual column stats within MT column_stats partition
+CREATE FUNCTION INDEX hour_of_day ON hudi_table USING BITMAP (DATE_FORMAT(timestamp, '%H')); -- functional index using bitmap for DATE_FORMAT(timestamp, '%H')
+CREATE FUNCTION INDEX income_idx ON employees (salary + (salary*commission_pct)); --  functional index using column stats for given expression
+-- NO CHANGE IN DROP INDEX --
+DROP INDEX last_name_idx;
+```
+
+`index_name` - Required, should be validated by parser. The name will be used for the partition name in MT.
+
+`index_type` - Optional, column\_stats if none provided and there are no functions and expressions in the command. Valid
```

Comment:
for col_stats, do we still have separate partitions? esp on the case where the `FUNCTION` keyword is not specified? Clarifying this would be great.

Reply:
yes, in the above scenario, we will have separate partition for `column_stats` as usual. I have an example in the sql code block above.
```
CREATE INDEX city_id_idx ON hudi_table (city_id); -- usual column stats within MT column_stats partition
```